MapleServer2/PacketHandlers/Game/BeautyHandler.cs

```diff
@@ -0,0 +1,611 @@
+                    break;
+                case BeautyMode.DyeItem:
+                    HandleDyeItem(session, packet);
+                    break;
+                default:
+                    IPacketHandler<GameSession>.LogUnknownMode(mode);
+                    break;
+            }
+        }
+
+        private static void HandleLoadShop(GameSession session, PacketReader packet)
+        {
+            int npcId = packet.ReadInt();
+            BeautyCategory category = (BeautyCategory) packet.ReadByte();
+
+            NpcMetadata beautyNpc = NpcMetadataStorage.GetNpc(npcId);
+            if (beautyNpc == null)
+            {
+                return;
+            }
+
+            BeautyMetadata beautyShop = BeautyMetadataStorage.GetShopById(beautyNpc.ShopId);
+            if (beautyShop == null)
+            {
+                return;
+            }
+
+            if (beautyShop.BeautyCategory == BeautyCategory.Dye)
+            {
+                if (beautyShop.BeautyType == BeautyShopType.Dye)
+                {
+                    session.Send(BeautyPacket.LoadDyeShop(beautyShop));
+                }
+                else
+                {
+                    session.Send(BeautyPacket.LoadBeautyShop(beautyShop));
+                }
+                return;
+            }
+
+            if (beautyShop.BeautyCategory == BeautyCategory.Save)
+            {
+                session.Send(BeautyPacket.LoadSaveShop(beautyShop));
+                session.Send(BeautyPacket.InitializeSaves());
+                session.Send(BeautyPacket.LoadSaveWindow());
+                session.Send(BeautyPacket.LoadSavedHairCount((short) session.Player.HairInventory.SavedHair.Count));
+                if (session.Player.HairInventory.SavedHair.Count != 0)
+                {
+                    session.Player.HairInventory.SavedHair = session.Player.HairInventory.SavedHair.OrderBy(o => o.CreationTime).ToList();
+                    session.Send(BeautyPacket.LoadSavedHairs(session.Player.HairInventory.SavedHair));
+                }
+
+                return;
+            }
+
+            List<BeautyItem> beautyItems = BeautyMetadataStorage.GetGenderItems(beautyShop.ShopId, session.Player.Gender);
+
+            session.Send(BeautyPacket.LoadBeautyShop(beautyShop, beautyItems));
+        }
+
+        private static void HandleNewBeauty(GameSession session, PacketReader packet)
+        {
+            byte unk = packet.ReadByte();
+            bool useVoucher = packet.ReadBool();
+            int beautyItemId = packet.ReadInt();
+            EquipColor equipColor = packet.Read<EquipColor>();
+            int colorIndex = packet.ReadInt();
+
+            Item beautyItem = new Item(beautyItemId) { Color = equipColor, IsTemplate = false };
+            BeautyMetadata beautyShop = BeautyMetadataStorage.GetCosmeticShopByItemId(beautyItem.Id);
+
+            if (useVoucher)
+            {
+                PayWithVoucher(session, beautyShop);
```

Comment:
In the method `PayWithVoucher` you can make it extend the same session, for simplicity and just pass the `beatuyShop`, see below on the method, same with `PayWithShopItemTokenCost`